metadata-service/plugin/src/main/java/com/datahub/plugins/loader/IsolatedClassLoaderImpl.java

```diff
@@ -0,0 +1,294 @@
+  }
+
+  private byte[] getClassData(ZipEntry zipEntry) {
+    try (InputStream ins = this._pluginJarRef.getInputStream(zipEntry);
+        ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      byte[] buffer = new byte[4096];
+      int bytesNumRead;
+      while ((bytesNumRead = ins.read(buffer)) != -1) {
+        baos.write(buffer, 0, bytesNumRead);
+      }
+      return baos.toByteArray();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return new byte[]{};
+  }
+
+  @Override
+  protected Class<?> loadClass(String s, boolean b) throws ClassNotFoundException {
+    log.debug(String.format("Load class %s", s));
+    String path = this.classNameToPath(s);
+    log.debug(String.format("File path %s", path));
+    // Check if requested class is available in plugin jar entries
+    if (!this._classPathVsZipEntry.containsKey(path)) {
+      // Try to load using Application class loader
+      log.debug(String.format("Class %s not found in plugin jar, trying application class loader chain", s));
+      for (ClassLoader classLoader : this._classLoaders) {
+        try {
+          log.debug("Looking in ClassLoader {}", classLoader.getClass().getName());
+          return classLoader.loadClass(s);
+        } catch (ClassNotFoundException classNotFoundException) {
+          // Pass it and let search in next ClassLoader
+        }
+      }
+      throw new ClassNotFoundException();
+    }
+
+    byte[] classBytes = getClassData(this._classPathVsZipEntry.get(path));
+    if (classBytes.length == 0) {
+      throw new ClassNotFoundException();
+    }
+    ProtectionDomain protectionDomain =
+        this._pluginPermissionManager.createProtectionDomain(this._pluginConfig.getPluginHomeDirectory());
+    return defineClass(s, classBytes, 0, classBytes.length, protectionDomain);
+  }
+
+  @Override
+  public URL getResource(String s) {
+    log.debug("Get resource {}", s);
+    return this.findResource(s);
+  }
+
+  @Override
+  public Enumeration<URL> getResources(String s) throws IOException {
+    URL url = this.getResource(s);
+    if (url == null) {
+      log.debug("Returning empty enumeration");
+      return Collections.emptyEnumeration();
+    }
+    List<URL> urls = new ArrayList<>(1);
+    urls.add(url);
+    return Collections.enumeration(urls);
+  }
+
+  @Override
+  public InputStream getResourceAsStream(String s) {
+    log.debug("Resource as stream = {}", s);
+    try {
+      URL url = this.findResource(s);
+      if (url == null) {
+        return null;
+      }
+      return url.openStream();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Look for resource in below order
+   * - First search in plugin jar if not found
+   * - then search in plugin directory if not found then return null
+   * @param resource Resource to find
+   * @return URL of the resource
+   */
+  @Override
+  protected URL findResource(String resource) {
```

Comment:
This method is too long. It can be broken down

Reply:
Done